File: web/src/workbench/directory-cache.test.mjs
import test from "node:test";
import assert from "node:assert/strict";
import { DirectoryCache } from "./directory-cache.ts";

function agent(revision, title = "自动标题", incarnation = "original") {
 const binding = { runner_id: "runner", fabric_id: "fabric", machine_id: "machine", revision: 1 };
 const runtime = { observation: { epoch: "connector", revision }, id: "runtime", incarnation, generation: 1, adapter: "acp", title: "default", session_metadata: { revision, conversation_id: "conversation", title } };
 return { agent_ref: `ref-${revision}`, target: { binding, runtime: { ...runtime } }, runtime };
}

test("final event wins delayed discovery above JavaScript safe integers, clear remains explicit", () => {
 const cache = new DirectoryCache(), batch = cache.begin("one");
 cache.apply(batch, { subscription_id: "one", kind: "member", agent: agent("9007199254740994") });
 cache.mergePage(batch, { items: [agent("9007199254740993", "older")] });
 assert.equal(cache.snapshot()[0].runtime.session_metadata.title, "自动标题");
 const absent = agent("9007199254740994");
 absent.runtime.observation.revision = "9007199254740995";
 absent.runtime.availability = "unavailable"; delete absent.runtime.session_metadata;
 cache.mergePage(batch, { items: [absent] });
 assert.equal(cache.snapshot()[0].runtime.session_metadata.revision, "9007199254740994");
 assert.equal(cache.snapshot()[0].agent_ref, "ref-9007199254740994");
 assert.equal(cache.snapshot()[0].runtime.availability, "unavailable");
 const cleared = agent("9007199254740995", null); cleared.runtime.observation.revision = "9007199254740996";
 cache.apply(batch, { subscription_id: "one", kind: "metadata", agent: cleared });
 assert.equal(cache.snapshot()[0].runtime.session_metadata.title, null);
});

for (const reason of ["removal", "replacement", "revocation"]) test(`${reason} invalidates already resolved pages and buffered events across resubscription`, () => {
 const cache = new DirectoryCache(), old = cache.begin("one");
 const held = agent("9007199254740999", "late higher title");
 cache.mergePage(old, { items: [held] });
 cache.apply(old, { subscription_id: "one", kind: "invalidated" });
 assert.equal(cache.mergePage(old, { items: [held] }), false);
 const current = cache.begin("two");
 cache.mergePage(current, { items: [agent("1", "replacement", "new")] });
 assert.equal(cache.apply(old, { subscription_id: "one", kind: "member", agent: held }), false);
 assert.equal(cache.mergePage(old, { items: [held] }), false);
 assert.equal(cache.apply(current, { subscription_id: "two", kind: "metadata", agent: held }), false);
 assert.equal(cache.snapshot().length, 1);
 assert.equal(cache.snapshot()[0].runtime.incarnation, "new");
});

for (const newerFromPage of [false, true]) test(`full Runtime ordering with unchanged title: ${newerFromPage ? "late event" : "late page"}`, () => {
 const cache = new DirectoryCache(), batch = cache.begin("one");
 const pending = agent("9007199254740994"); pending.runtime.state = "running";
 const confirmed = structuredClone(pending);
 confirmed.runtime.observation.revision = "9007199254740995";
 confirmed.agent_ref = "confirmed-reference";
 confirmed.runtime.native_session = { id: "native", cwd: "/workspace" };
 const exited = structuredClone(confirmed);
 exited.runtime.observation.revision = "9007199254740996";
 Object.assign(exited.runtime, { state: "exited", exit_code: 0, stop_reason: "completed" });
 const page = value => cache.mergePage(batch, { items: [value] });
 const event = value => cache.apply(batch, { subscription_id: "one", kind: "member", agent: value });
 const accept = newerFromPage ? page : event, delayed = newerFromPage ? event : page;
 for (const value of [pending, confirmed, exited]) { assert.equal(accept(value), true); assert.deepEqual(cache.snapshot(), [value]); }
 for (const value of [pending, confirmed, exited]) { assert.equal(delayed(value), true); assert.deepEqual(cache.snapshot(), [exited]); }
});

test("observation epoch changes require a fresh subscription and fence old discovery", () => {
 const cache = new DirectoryCache(), batch = cache.begin("one"), old = agent("100");
 cache.mergePage(batch, { items: [old] });
 const current = structuredClone(old); current.runtime.observation = { epoch: "reconnected", revision: "1" };
 assert.equal(cache.apply(batch, { subscription_id: "one", kind: "member", agent: current }), false);
 assert.deepEqual(cache.snapshot(), []);
 const fresh = cache.begin("two");
 assert.equal(cache.mergePage(fresh, { items: [current] }), true);
 assert.equal(cache.mergePage(batch, { items: [old] }), false);
 assert.deepEqual(cache.snapshot(), [current]);
});

File: pkg/agents/directory_cache_test.go
package agents

import (
	"testing"

	"github.com/aiomni/dune/pkg/api"
	"github.com/aiomni/dune/pkg/runner"
	"github.com/aiomni/dune/pkg/workbench"
)

func directoryCacheAgent(revision uint64, title string) Agent {
	binding := runner.Binding{RunnerID: "runner", FabricID: "fabric", MachineID: "machine", Revision: 1}
	conversation := "conversation"
	runtime := api.Runtime{ID: "runtime", Incarnation: "original", Generation: 1, Adapter: "acp", ConversationID: conversation,
		SessionMetadata: &api.SessionMetadata{Revision: revision, ConversationID: &conversation, Title: &title}}
	return Agent{Target: workbench.AgentTarget{Binding: binding, Runtime: workbench.RuntimeRef{ID: runtime.ID, Incarnation: runtime.Incarnation, Generation: runtime.Generation, Adapter: runtime.Adapter}},
		Runtime: runtime, Runner: runner.Runner{ID: binding.RunnerID, Binding: &binding}}
}

func TestDirectoryCacheFencesBufferedPagesAndEventsAfterInvalidation(t *testing.T) {
	for _, reason := range []string{"removed", "replaced", "revoked"} {
		t.Run(reason, func(t *testing.T) {
			var cache DirectoryCache
			batch := cache.Begin("original-subscription")
			old := directoryCacheAgent(1<<53, "old")
			if !cache.MergePage(batch, DirectoryPage{Items: []Agent{old}}) {
				t.Fatal("initial discovery rejected")
			}
			cache.Apply(DirectoryEvent{SubscriptionID: "original-subscription", Kind: DirectoryInvalidated, Code: reason})
			late := directoryCacheAgent(1<<53+100, "late but higher")
			page := DirectoryPage{Items: []Agent{late}}
			if cache.MergePage(batch, page) || cache.Apply(DirectoryEvent{SubscriptionID: "original-subscription", Kind: DirectoryMember, Agent: &late}) || len(cache.Snapshot()) != 0 {
				t.Fatal("invalidated member revived")
			}
			current := cache.Begin("new-subscription")
			if cache.MergePage(batch, page) || cache.Apply(DirectoryEvent{SubscriptionID: "original-subscription", Kind: DirectoryMember, Agent: &late}) {
				t.Fatal("reconnecting relabeled old data")
			}
			if cache.Apply(DirectoryEvent{SubscriptionID: "new-subscription", Kind: DirectoryMetadata, Agent: &late}) {
				t.Fatal("title alone established membership")
			}
			confirmed := directoryCacheAgent(1, "confirmed replacement")
			confirmed.Runtime.Incarnation, confirmed.Target.Runtime.Incarnation = "replacement", "replacement"
			if !cache.MergePage(current, DirectoryPage{Items: []Agent{confirmed}}) || len(cache.Snapshot()) != 1 || cache.Snapshot()[0].Runtime.Incarnation != "replacement" {
				t.Fatal("newly confirmed member was not accepted")
			}
		})
	}
}

func TestDirectoryCacheMergesSingleRevisionAfterMembership(t *testing.T) {
	var cache DirectoryCache
	batch := cache.Begin("subscription")
	newer := directoryCacheAgent(1<<53+2, "new")
	newer.Ref = "current-native-reference"
	if !cache.Apply(DirectoryEvent{SubscriptionID: "subscription", Kind: DirectoryMember, Agent: &newer}) {
		t.Fatal("member event rejected")
	}
	older := directoryCacheAgent(1<<53+1, "old")
	older.Ref = "old-native-reference"
	cache.MergePage(batch, DirectoryPage{Items: []Agent{older}})
	absent := older
	absent.Runtime.SessionMetadata = nil
	cache.MergePage(batch, DirectoryPage{Items: []Agent{absent}, Complete: false})
	if value := cache.Snapshot()[0].Runtime.SessionMetadata; value.Revision != newer.Runtime.SessionMetadata.Revision || *value.Title != "new" {
		t.Fatal("late or unavailable discovery overwrote a newer title", value)
	}
	if cache.Snapshot()[0].Ref != newer.Ref {
		t.Fatal("old discovery replaced the current native action reference")
	}
	cleared := directoryCacheAgent(1<<53+3, "")
	cleared.Runtime.SessionMetadata.Title = nil
	cache.Apply(DirectoryEvent{SubscriptionID: "subscription", Kind: DirectoryMetadata, Agent: &cleared})
	if cache.Snapshot()[0].Runtime.SessionMetadata.Title != nil {
		t.Fatal("explicit clear was lost")
	}
	// Neither caller input nor a returned snapshot may mutate retained data.
	cleared.Runtime.SessionMetadata.Revision = 1
	copy := cache.Snapshot()
	copy[0].Runtime.SessionMetadata.Revision = 2
	if cache.Snapshot()[0].Runtime.SessionMetadata.Revision != 1<<53+3 {
		t.Fatal("mutable metadata escaped the SDK cache")
	}
}
